### PR TITLE
[ci] Build pipeline optimizing

### PR DESCRIPTION
**Pull request by apolovov**
## Description
Build optimizations for the CI pipeline:
* modules_images and deckhouse builds were united to a single werf.yaml. No more separate modules_images build and awkward dependence on cached images_tags files.
* Common go pkg cache for go-powered images was implemented.
* Some legacy binaries were excluded from the deckhouse image (rgxg, apache2-utils, netcat-openbsd).
* Artifact importings were moved to "after setup" stage instead of "before setup" to reduce unwanted rebuilding.
* Lot of files were excluded from building artifacts. I.e. "_test.go" ones (they used to trigger unwanted rebuilding for deckhouse-controller image).
* Fixed bug in build.yaml ci template — the 'dev' image never deleted from a runner after building/pushing. Only 'dev/install' did.
* Got rid of /dev suffix for dev branches. It used to distinguish images in a single registry. Now we use only tags.
* Go Generate step was moved to our runners from GitHub ones. So we can cache docker image and go pkg now.
* After-building push (for human-readable tag) was optimized by getting rid of slow `werf stage image dev`.

Affected images to rebuild and restart related Pods:
```
007-registrypackages/containerd-fe
021-kube-proxy/kube-proxy
030-cloud-provider-aws/cloud-controller-manager
030-cloud-provider-gcp/cloud-controller-manager
030-cloud-provider-openstack/cinder-csi-plugin
030-cloud-provider-openstack/cloud-controller-manager
030-cloud-provider-vsphere/cloud-controller-manager
030-cloud-provider-yandex/cloud-controller-manager
040-control-plane-manager/control-plane-manager
040-control-plane-manager/kube-apiserver
040-control-plane-manager/kube-apiserver-healthcheck
040-control-plane-manager/kube-controller-manager
040-node-manager/bashible-apiserver
040-node-manager/machine-controller-manager
040-terraform-manager/base-terraform-manager
040-terraform-manager/terraform-manager-gcp
040-terraform-manager/terraform-manager-vsphere
```

## Why do we need it, and what problem does it solve?
No one loves wasting his life watching the black screen with jumping symbols. We are no exception.

## Changelog entries

```changes
module: ci
type: feature
description: Build optimizations for the CI pipeline.
note: Some pods including control-plane will be restarted. ingress-nginx is not one of them.
```

## Before merge

- [x] Ask @diafour to revert temporary github workflow fix (on push...).
- [x] Ask @z9r5 to check web-related stuff.
- [x] Ask @z9r5 to check release-related stuff.
- [x] Measure the build speed gain by rebasing some existing PRs.
- [x] Try to setup a cluster from scratch.
- [x] Optimize [generator](https://github.com/deckhouse/deckhouse/pull/616/files#diff-7e9daa94d6bf20cc7735c2e8d0e0f23503bbf180fa4d06b078ce821840acebd6R56). And handle 007-registrypackages no-hooks problem (i-am-hack-delete-me file). I've opened [an issue](https://github.com/deckhouse/deckhouse/issues/895).
- [x] Make decision about "Go generate" job dependencies.
- [x] Cache go pkg for "Go Generate". 
- [x] Exclude everything from image: dhctl-tests.
- [x] Discover better werf parallelism --parallel-tasks-limit=5.
- [x] Add go pkg cache for tests running.
- [x] do something with .gitlab-ci-simple.yml.
- [x] `werf stage image dev` and `werf stage image dev/install` work too long
- [x] 404 to bb-rp-install
- [x] get rid of /dev path (don't forget about e2e an tfcloud.sh)

### CODE REVIEW

**Comment by apolovov** [comment]
Optimization needed:
https://github.com/deckhouse/deckhouse/pull/616/files#diff-7e9daa94d6bf20cc7735c2e8d0e0f23503bbf180fa4d06b078ce821840acebd6R56

File: werf.yaml [hunk 1]
```diff
@@ -744,3 +675,121 @@ shell:
...
+  {{- $_ := set $ctx "ModulePath" $ImageData._0 }}
+  {{- $_ := set $ctx "ModulePriority" $ImageData._1 }}
+  {{- $_ := set $ctx "ModuleName" $ImageData._2 }}
+  {{- $_ := set $ctx "ImageName" $ImageData._3 }}
+  {{- $_ := set $ctx "ImageBuildData" $content }}
+  {{- $_ := set $ctx "Files" $Root.Files }}
+  {{- $_ := set $ctx "Env" $Root.Env }}
+---
+{{ include "module_image_template" $ctx }}
+  {{- range $ImageYamlMainfest := (include "module_image_template" $ctx | splitList "---\n") }}
```
Comment (diafour) [inline werf.yaml:739]: "---\n" is not a safe separator for the split. Imagine someone decides to embed an SSL certificate in the image. "^---\n" is safer, but we can't use it here. So just a note.

File: werf.yaml [hunk 2]
```diff
@@ -744,3 +675,121 @@ shell:
...
+  - |
+    env | grep MODULES_IMAGES_TAGS | jq -Rn '
+      reduce inputs as $i (
+        {};
+        . * (
+          $i | ltrimstr("MODULES_IMAGES_TAGS_") | sub("=";"_") |
+          split("_")[0] as $moduleName |
+          split("_")[1] as $imageName |
+          split("_")[2] as $tag |
+          {($moduleName): {($imageName): $tag}}
```
Comment (diafour) [inline werf.yaml:775]:
```suggestion
          $i | ltrimstr("MODULES_IMAGES_TAGS_") | sub("=";"_") |
          split("_") as [$moduleName, $imageName, $tag] |
          {($moduleName): {($imageName): $tag}}
```

WDYT?

**Review by diafour — CHANGES_REQUESTED** [review]
Looks good. I have only minor concerns.

**Comment by diafour** [comment]
One more thought: we should change Gitlab configuration. At least everything related to .gitlab-ci-simple.yml

File: .github/ci_templates/build.yml [hunk 3]
```diff
@@ -145,28 +115,37 @@ steps:
...
-          pull_push 'release-channel-version' $(werf stage image release-channel-version) ${DESTINATION_RELEASE_CHANNEL_VERSION_IMAGE}
+          RELEASE_CHANNEL_VERSION_IMAGE_URL="$(jq -r '.Images."release-channel-version".DockerImageName' images_tags_werf.json)"
+          pull_push 'release-channel-version' ${RELEASE_CHANNEL_VERSION_IMAGE_URL} ${DESTINATION_RELEASE_CHANNEL_VERSION_IMAGE}
 
-          echo Remove local tags.
+          echo "⚓️ [$(date -u)] Remove local tags."
           docker image rmi ${DECKHOUSE_DESTINATION_IMAGE} || true
           docker image rmi ${DECKHOUSE_DESTINATION_INSTALL_IMAGE} || true
-          docker image rmi ${DESTINATION_RELEASE_CHANNEL_VERSION_IMAGE} || true
+          docker image rmi ${DESTINATION_RELEASE_CHANNEL_VERSION_IMAGE} || truea
```
Comment (diafour) [inline .github/ci_templates/build.yml:144]:
```suggestion
          docker image rmi ${DESTINATION_RELEASE_CHANNEL_VERSION_IMAGE} || true
```

**Review by diafour — CHANGES_REQUESTED** [review]

**Review by nabokihms — APPROVED** [review]

**Pull request merged**